persist/fs/bloom_filter.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fs
+
+import (
+	"os"
+
+	"github.com/m3db/bloom"
+	"github.com/m3db/m3db/digest"
+)
+
+type bloomFilter interface {
+	// Test whether the Bloom Filter contains a value
+	Test(value []byte) bool
+
+	// Returns the number of elements in the Bloom Filter
+	M() uint
+
+	// Returns the number of hashes used
+	K() uint
+}
+
+// managedBloomFilter is a container object that implements lifecycle
+// management ontop of a BloomFilter. I.E it wraps a bloom filter such that
+// all resources are released when the Close() method is called
+type managedBloomFilter interface {
+	bloomFilter
+
+	// Close closes the ManagedBloomFilter, releasing any held resoures
+	Close() error
+}
+
+// managedConcurrentBloomFilter is the same as managedBloomFilter, with the
+// addition that the implementation must be safe for concurrent use
+type managedConcurrentBloomFilter interface {
+	managedBloomFilter
+
+	IsConcurrent()
```

Comment:
Haha - I like where you went here, forcing Go's type system to be useful can be challenging sometimes.  

Considering all these types are internal to just this package it might be simpler to just make a single struct `type managedConcurrentBloomFilter struct { ... }` (there won't be a no need to mock these I don't think, so the value of an interface is not as high).

Reply:
Yeah its a cool little pattern I've seen in some open-source code before. I think you're right though, I probably overdid it in this case. I'll just make it the one struct